yoyodyne/models/hmm.py

```diff
@@ -0,0 +1,413 @@
+            predictions, self.end_idx, self.pad_idx
+        )
+        val_eval_item = self.evaluator.get_eval_item(
+            predictions, batch.target.padded, self.pad_idx
+        )
+        return {"val_eval_item": val_eval_item, "val_loss": -likelihood.mean()}
+
+    def predict_step(self, batch: data.PaddedBatch, batch_idx: int) -> Dict:
+        predictions, _ = self(batch)
+        # Processes for accuracy calculation.
+        predictions = self.evaluator.finalize_predictions(
+            predictions, self.end_idx, self.pad_idx
+        )
+        return predictions
+
+    def get_decoder(self) -> modules.lstm.HMMLSTMDecoder:
+        if self.attention_context > 0:
+            return modules.lstm.ContextHMMLSTMDecoder(
+                pad_idx=self.pad_idx,
+                start_idx=self.start_idx,
+                end_idx=self.end_idx,
+                attention_context=self.attention_context,
+                decoder_input_size=(
+                    self.source_encoder.output_size
+                    + self.features_encoder.output_size
+                    if self.has_features_encoder
+                    else self.source_encoder.output_size
+                ),
+                num_embeddings=self.target_vocab_size,
+                dropout=self.dropout,
+                bidirectional=False,
+                embedding_size=self.embedding_size,
+                layers=self.decoder_layers,
+                hidden_size=self.hidden_size,
+            )
+        else:
+            return modules.lstm.HMMLSTMDecoder(
+                pad_idx=self.pad_idx,
+                start_idx=self.start_idx,
+                end_idx=self.end_idx,
+                decoder_input_size=(
+                    self.source_encoder.output_size
+                    + self.features_encoder.output_size
+                    if self.has_features_encoder
+                    else self.source_encoder.output_size
+                ),
+                num_embeddings=self.target_vocab_size,
+                dropout=self.dropout,
+                bidirectional=False,
+                embedding_size=self.embedding_size,
+                layers=self.decoder_layers,
+                hidden_size=self.hidden_size,
+            )
+
+    def _get_loss_func(
+        self,
+    ) -> Callable[[torch.Tensor, torch.Tensor, torch.Tensor], torch.Tensor]:
+        # TODO: Make the loss update continuous during training to avoid.
+        # Also enable better reflection of val loss.
+        """Returns the actual function used to compute loss.
+
+        Returns:
+                Callable[[torch.Tensor, torch.Tensor], torch.Tensor]:
+                    configured loss function.
+        """
+        return self.loss
+
+    def loss(self, target, log_probs, alignments):
+        fwd = alignments[0, :, 0].unsqueeze(1) + self._gather_at_idx(
+            log_probs[0], target[:, 0]
+        )
+        for t in range(1, target.shape[1]):
```

Comment:
more informative variable name?

Reply:
I feel like this is pretty conventional. It's a decode function for LSTM. We're pretty consistent with using 1 indexing for length.

Reply:
Well, what is `t`? "time"?

Reply:
t for each t step in target.